### PR TITLE
Bug/digitizing z coord

### LINKED ISSUE #286: Does Input take Z values

### PR DESCRIPTION
**Pull request by sklencar**
Z coord recording is activated either immediately when entering record mode or after clicking on GPS icon on record toolbar. It is disabled after manual move of extent.
Crosshair is sticked with GPS marker while z coord recording is enabled.

closes #286

### CODE REVIEW

**Comment by jancelin** [comment]
Hi!
Thanks for this dev
But Input crash when I add a New pointZ (postgis connexion)....
I Can send you a qgis project by mail if you want.  julien.ancelin@inrae.fr

**Comment by saberraz** [comment]
Could you add me to your  project?
My username for Mergin is saber

**Comment by jancelin** [comment]
@saberraz  I just sent you the files

**Comment by jancelin** [comment]
Hi

This latest version does not install.

 > signed apk: [arm64_v8a](https://www.dropbox.com/s/8m7vrq900ccbipy/input-588-c1f6852f9c40ad3f98e6f592b5f18f466bc4ba17-arm64_v8a-signed.apk?dl=0) (SDK: [android-5](https://github.com/lutraconsulting/input-sdk/releases/tag/android-5))
This latest version does not install.

 The previous one works.

https://github.com/lutraconsulting/input/pull/588#issuecomment-596523437

**Comment by jancelin** [comment]
And adding Z to a PointZ works well. 👍
Thanks

**Comment by jancelin** [comment]
I can have enough decimal for the Z 
Screenshot_20200310-185840

**Comment by jancelin** [comment]
Is it possible to have a working build?

> signed apk: [arm64_v8a](https://www.dropbox.com/s/8m7vrq900ccbipy/input-588-c1f6852f9c40ad3f98e6f592b5f18f466bc4ba17-arm64_v8a-signed.apk?dl=0) (SDK: [android-5](https://github.com/lutraconsulting/input-sdk/releases/tag/android-5))

**Comment by sklencar** [comment]
Hi @jancelin, good news, the builds have passed. Many thanks for your interest and remarks.

**Comment by jancelin** [comment]
Hi @sklencar, it's working !
Thanks.

File: app/digitizingcontroller.cpp [hunk 1]
```diff
@@ -200,11 +244,7 @@ void DigitizingController::onPositionChanged()
     return;
 
   QgsPoint point = mPositionKit->position();
-  QgsGeometry geom( point.clone() );
-  geom.transform( transformer() );
-
-  QgsPoint *layerPoint = qgsgeometry_cast<QgsPoint *>( geom.get() );
-  fixZ( layerPoint );
+  QgsPoint *layerPoint = getLayerPoint( point, true );
```
Comment (wonder-sk) [inline app/digitizingcontroller.cpp]: layerPoint here is not getting deleted... maybe better to return std::unique_ptr<QgsPoint> from getLayerPoint() to make things safer ?

File: app/digitizingcontroller.cpp [hunk 2]
```diff
@@ -152,27 +163,60 @@ void DigitizingController::setLineRecordingInterval( int lineRecordingInterval )
...
 
+QgsQuickFeatureLayerPair DigitizingController::pointFeatureFromPoint( const QgsPoint &point, bool isGpsPoint )
+{
+  if ( !featureLayerPair().layer() )
+    return QgsQuickFeatureLayerPair();
+
+  if ( !mMapSettings )
+  {
+    return QgsQuickFeatureLayerPair();
+  }
```
Comment (wonder-sk) [inline app/digitizingcontroller.cpp]: this could be added to the previous if() to make the code more concise

File: app/digitizingcontroller.cpp [hunk 3]
```diff
@@ -261,26 +298,20 @@ QgsPoint DigitizingController::pointFeatureMapCoordinates( QgsQuickFeatureLayerP
...
+void DigitizingController::addRecordPoint( const QgsPoint &point, bool isGpsPoint )
 {
   if ( !mRecording )
     return;
 
-  QgsPointXY layerPointXY = mMapSettings->mapSettings().mapToLayerCoordinates( featureLayerPair().layer(), QgsPointXY( point.x(), point.y() ) );
-  QgsPoint layerPoint( layerPointXY );
-  fixZ( &layerPoint );
-  mRecordedPoints.append( layerPoint );
+  QgsPoint *layerPoint = getLayerPoint( point, isGpsPoint );
```
Comment (wonder-sk) [inline app/digitizingcontroller.cpp]: same here - leaking layerPoint

File: app/qml/main.qml [hunk 4]
```diff
@@ -314,11 +319,10 @@ ApplicationWindow {
...
 
       onScreenPositionChanged: {
-        if (__appSettings.autoCenterMapChecked) {
-          var border = mainPanel.height
-          if (isPositionOutOfExtent(border)) {
+        if (digitizing.useGpsPoint || (__appSettings.autoCenterMapChecked && isPositionOutOfExtent(mainPanel.height))) {
+            var useGpsPoint = digitizing.useGpsPoint
             mapCanvas.mapSettings.setCenter(positionKit.projectedPosition);
-          }
+            digitizing.useGpsPoint = useGpsPoint
```
Comment (wonder-sk) [inline app/qml/main.qml]: I am not sure if I understand what is going on here and why

File: app/qml/main.qml [hunk 5]
```diff
@@ -84,9 +84,20 @@ ApplicationWindow {
             popup.open()
         }
         stateManager.state = "view"
+        digitizing.useGpsPoint = false
```
Comment (wonder-sk) [inline app/qml/main.qml]: Not sure why we are setting this here... maybe worth a comment?

**Comment by sklencar** [comment]
Note: Currently, there is no manual switch (button or anything like that) for changing non/GPS recorded point type (thus if recorded point contains z-coord from GPS). Implementation has simple logic - if you clicked on gps icon, record cross marker will follow gps marker -> recorded point is GPS point and contains z-coord. By any move of canvas (pan, zoom in/out) switches recorded mode to regular mode, thus without z-coord.

File: app/digitizingcontroller.cpp [hunk 6]
```diff
@@ -87,6 +87,11 @@ void DigitizingController::fixZ( QgsPoint *point ) const
   bool layerIs3D = QgsWkbTypes::hasZ( featureLayerPair().layer()->wkbType() );
   bool pointIs3D = QgsWkbTypes::hasZ( point->wkbType() );
 
+
+  QgsWkbTypes::GeometryType faeetureType = featureLayerPair().feature().geometry().type();
+  QgsWkbTypes::GeometryType layerType = featureLayerPair().layer()->geometryType();
+
+  qDebug() << "!!!!!COMPATE TYPES !!!!" << faeetureType << layerType;
```
Comment (wonder-sk) [inline app/digitizingcontroller.cpp]: Can be removed?

File: app/digitizingcontroller.cpp [hunk 7]
```diff
@@ -77,7 +77,7 @@ bool DigitizingController::isPairValid( QgsQuickFeatureLayerPair pair ) const
   return pair.isValid() && hasEnoughPoints();
 }
 
-void DigitizingController::fixZ( QgsPoint *point ) const
+void DigitizingController::fixZ( std::unique_ptr<QgsPoint> &point ) const
```
Comment (wonder-sk) [inline app/digitizingcontroller.cpp]: can we simplify this to `QgsPoint &point` - there's no change of ownership so best to just pass a non-const reference

**Pull request merged**